nix/opam2nix.nix

```diff
@@ -0,0 +1 @@
+import (builtins.fetchTarball "https://github.com/timbertson/opam2nix/archive/v1.tar.gz") {}
```

Comment:
Shouldn't we specify a sha256 here? In fact I'm surprised it's even possible to omit it.

Alternatively, if the purpose of this file is a rare manual impure operation that updates `opam-selection.nix`, then maybe we should rename this to `opam2nix-impure.nix` and then we can be careful when calling impure stuff. (Is there a nix flag that forbids impure computations like this? If there is, I want that to be the default)

By the way, it would be nice to have comments explaining in short how to use the nix workflow (maybe at the top of shell.nix?).

Reply:
I'm not that familiar with nix, so feel free to correct me here, but is this any worse than importing nixpkgs? Nixpkgs is also different between runs if one updates their channels.

> then maybe we should rename this to opam2nix-impure.nix and then we can be careful when calling impure stuff

Alternatively, we could specify a git commit? This would prevent the source from changing implicitly.

> (Is there a nix flag that forbids impure computations like this? If there is, I want that to be the default)

The only thing that I'm aware of is the experimental "flakes" features which allows to lock all nix sources to make it completely reproducible.

> By the way, it would be nice to have comments explaining in short how to use the nix workflow (maybe at the top of shell.nix?).

Will do. Everything is standard nix stuff unless one wants to regenerate the build plan using opam2nix.

Reply:
Things may have changed since I actively used nix, but my understanding was that only the filesystem could be read by the nix interpreter, and the network access had to be guarded by sha256. So what I see here definitely looks weird with my, perhaps old-fashioned, understanding of nix.

> is this any worse than importing nixpkgs?

Nixpkgs is (normally) imported from a local directory, which needs to be updated separately. Nix language evaluation is not involved in that.

> we could specify a git commit?

Sure, but that's not enough to create a fixed-output nix derivation, you have to convince nix, and SHA1 is not enough to convince it. Again, I don't understand why this works as written, I think it shouldn't. But if it does work, then fine.

> the experimental "flakes" features

I've been meaning to read about them for a while, but couldn't find the time. It does sound interesting.

Reply:
> Sure, but that's not enough to create a fixed-output nix derivation, you have to convince nix, and SHA1 is not enough to convince it. Again, I don't understand why this works as written, I think it shouldn't. But if it does work, then fine.

I confirmed in the docs for tarball and the sha is optional. In any case, I changed it to fetchGit + a `rev`. Seems a little safer and we don't want opam2nix to randomly change for us.

Reply:
I see, this `fetchGit` is indeed impure and is done at evaluation time similar to importing a file. It's a feature added in Nix 2.0, and I was thinking of the old pre-2.0 way to fetch things.

Looks like the you can tweak nix purity with `--pure-eval` or `--restrict-eval`.